src/utils/workflow_utils/src/workflow_utils.c

```diff
@@ -1804,57 +1804,86 @@ bool workflow_set_sandbox(ADUC_WorkflowHandle handle, const char* sandbox)
     return true;
 }
 
-// Workfolder =  [root.sandboxfolder]  "/"  ( [parent.workfolder | parent.id]  "/" )+  [handle.workfolder | handle.id]
-char* workflow_get_workfolder(const ADUC_WorkflowHandle handle)
+// Returns the download base directory
+char* workflow_get_workflow_base_dir(const ADUC_WorkflowHandle handle)
 {
     char dir[1024] = { 0 };
-
-    char* pwf = NULL;
     char* ret = NULL;
-    char* id = workflow_get_id(handle);
-
-    // If workfolder explicitly specified, use it.
-    char* wf = workflow_get_string_property(handle, WORKFLOW_PROPERTY_FIELD_WORKFOLDER);
-    if (wf != NULL)
-    {
-        Log_Debug("Property '%s' not NULL - returning cached workfolder '%s'", WORKFLOW_PROPERTY_FIELD_WORKFOLDER, wf);
-        free(id);
-        return wf;
-    }
+    char* pwf = NULL;
 
-    // Return ([parent's workfolder] or [default sandbox folder]) + "/" + [workflow id];
     ADUC_WorkflowHandle p = workflow_get_parent(handle);
     if (p != NULL)
     {
         pwf = workflow_get_workfolder(p);
-        sprintf(dir, "%s/%s", pwf, id);
+        sprintf(dir, "%s/", pwf);
 
-        Log_Debug("Using parent workfolder: '%s/%s'", pwf, id);
+        Log_Debug("Using parent workfolder: '%s/%s'", pwf);
     }
     else
     {
         const ADUC_ConfigInfo* config = ADUC_ConfigInfo_GetInstance();
         if (config != NULL)
         {
-            Log_Info("Sandbox root path not set. Use default: '%s'", config->downloadsFolder);
-            sprintf(dir, "%s/%s", config->downloadsFolder, id);
+            Log_Info("No parent workfolder. Using default: '%s'", config->downloadsFolder);
+            sprintf(dir, "%s", config->downloadsFolder);
             ADUC_ConfigInfo_ReleaseInstance(config);
         }
         else
         {
             Log_Warn("Config is null. Use default sandbox root path: '%s'", DEFAULT_SANDBOX_ROOT_PATH);
-            sprintf(dir, "%s/%s", DEFAULT_SANDBOX_ROOT_PATH, id);
+            sprintf(dir, "%s", DEFAULT_SANDBOX_ROOT_PATH);
         }
     }
 
-    free(pwf);
-    free(id);
+    if (dir[0] != 0)
+    {
+        mallocAndStrcpy_s(&ret, dir);
+    }
+
+    return ret;
+}
+
+// Workfolder =  [root.sandboxfolder]  "/"  ( [parent.workfolder | parent.id]  "/" )+  [handle.workfolder | handle.id]
+char* workflow_get_workfolder(const ADUC_WorkflowHandle handle)
+{
+    char dir[1024] = { 0 };
+
+    char* bdir = NULL;
+    char* ret = NULL;
+    char* id = workflow_get_id(handle);
+
+    if (id == NULL)
+    {
+        return NULL;
+    }
+
+    // If workfolder explicitly specified, use it.
+    char* wf = workflow_get_string_property(handle, WORKFLOW_PROPERTY_FIELD_WORKFOLDER);
+    if (wf != NULL)
+    {
+        Log_Debug("Property '%s' not NULL - returning cached workfolder '%s'", WORKFLOW_PROPERTY_FIELD_WORKFOLDER, wf);
+        free(id);
```

Comment:
person who wrote this was against "goto done"?  A bunch of redundant free(id) calls.